common/sgx/internal.edl

```diff
@@ -11,17 +11,126 @@
+    untrusted
+    {
+        uint32_t oe_get_qetarget_info_ocall(
+            [out] sgx_target_info_t* target_info);
+
+        uint32_t oe_get_quote_ocall(
+            [in] const sgx_report_t* sgx_report,
+            [out, size=quote_size] void* quote,
+            size_t quote_size,
+            [out] size_t* quote_size_out);
+
+        uint32_t oe_get_qe_identify_info_ocall(
+            [out, size=qe_id_info_size] void* qe_id_info,
+            size_t qe_id_info_size,
+            [out, count=1] size_t* qe_id_info_size_out,
+            [out, size=issuer_chain_size] void* issuer_chain,
+            size_t issuer_chain_size,
+            [out, count=1] size_t* issuer_chain_size_out);
+
+        uint32_t oe_get_revocation_info_ocall(
+            [in] uint8_t fmspc[6],
+            size_t num_crl_urls,
+            [in, string] const char* crl_urls0,
+            [in, string] const char* crl_urls1,
+            [in, string] const char* crl_urls2,
+            [out, size=tcb_info_size] void* tcb_info,
+            size_t tcb_info_size,
+            [out, count=1] size_t* tcb_info_size_out,
+            [out, size=tcb_issuer_chain_size] void* tcb_issuer_chain,
+            size_t tcb_issuer_chain_size,
+            [out, count=1] size_t* tcb_issuer_chain_size_out,
+            [out, size=crl0_size] void* crl0,
+            size_t crl0_size,
+            [out, count=1] size_t* crl0_size_out,
+            [out, size=crl1_size] void* crl1,
+            size_t crl1_size,
+            [out, count=1] size_t* crl1_size_out,
+            [out, size=crl2_size] void* crl2,
+            size_t crl2_size,
+            [out, count=1] size_t* crl2_size_out,
+            [out, size=crl_issuer_chain0_size] void* crl_issuer_chain0,
+            size_t crl_issuer_chain0_size,
+            [out, count=1] size_t* crl_issuer_chain0_size_out,
+            [out, size=crl_issuer_chain1_size] void* crl_issuer_chain1,
+            size_t crl_issuer_chain1_size,
+            [out, count=1] size_t* crl_issuer_chain1_size_out,
+            [out, size=crl_issuer_chain2_size] void* crl_issuer_chain2,
+            size_t crl_issuer_chain2_size,
+            [out, count=1] size_t* crl_issuer_chain2_size_out);
+
+        void oe_log_ocall(
+            uint32_t log_level,
+            [in, string] const char* message);
+
+        // Translate the addresses in buffer[] into symbol names.
+        // (similar to Linux backtrace_symbols().
+        uint32_t oe_backtrace_symbols_ocall(
+            [user_check] oe_enclave_t* oe_enclave,
+            [in, count=size] const uint64_t* buffer,
+            size_t size,
+            [out, size=strings_buf_size] void* strings_buf,
+            size_t strings_buf_size,
+            [out, count=1] size_t* strings_buf_size_out);
+
+        void oe_thread_wake_wait_ocall(
+            [user_check] oe_enclave_t* oe_enclave,
+            uint64_t waiter_tcs,
+            uint64_t self_tcs);
+
+        void* oe_realloc_ocall(
```

Comment:
It would be preferable to leave realloc as a core OCALL. 
For SGX, Ocalls require host memory allocation just to make ocalls and this would mean that in order to do reallocation, we end up doing an allocation of marshaling struct in host memory and then freeing it.

Reply:
I thought so too until I realized that the original function has to allocate and free the following structure.

```c++
typedef struct _oe_realloc_args
{
    void* ptr;
    size_t size;
} oe_realloc_args_t;
```

The generated **oe_realloc_ocall** routines are just as efficient.

I added issue #2046 which proposes adding one or more additional arguments to **oe_ocall** for such situations. Also, the edger8r could use these additional arguments for optimization.

The **oe_ocall** and **oe_ecall** functions must carry simple 64-bit values (not pointers to host memory) because the TrustZone work needs to obtain host memory as well.

Reply:
Thanks for creating the issue.